### PR TITLE
gl/vk: Enable vertex texture fetch

### PR DESCRIPTION
**Pull request by kd-11**
Enable vertex textures.
Most of the rsx side was already done, so adding to the renderer should not be a big deal.
However, currently, the decompiler only addresses vertex texture 0.

This PR enables vertex texture fetch in GL backend only (for now)

### CODE REVIEW

**Comment by kd-11** [comment]
@raven02 Please check for regressions. The vertex texture sample works, but the texture seems to have repeat set to false (atleast for the W32_Z32_Y32_X32_FLOAT test) which confines the geometry to a small block. Maybe we shouldn't honor clamping modes?

**Comment by kd-11** [comment]
image

Sort of working now. Some bugs are visible but they seem unrelated, I think the strange geometry over the "terrain" may be a shader effect for depth fog that isn't working properly. If you can get to the underside of the "terrain" you can see the vertex deformations are correct.

**Comment by kd-11** [comment]
Spurious build failure with appveyor.

**Comment by raven02** [comment]
Nice @kd-11

**Comment by Nezarn** [comment]
Hatsune Miku Dreamy Theater, 2nd, and extend can now boot
http://i.imgur.com/LTAHxP1.png

ps: someone should restart appveyor, since it didn't had any build errors (at least i can't see any), and also VS builds fine

**Comment by raven02** [comment]
I tested a bit as well and looks okay for most games.

Appveyor build - https://ci.appveyor.com/project/raven02/rpcs3-ppr74/build/1.0.74/artifacts

**Comment by kd-11** [comment]
Thats good news. Will update with vulkan and dx12 support.

**Comment by kd-11** [comment]
Fixed opengl vertex issue, it seems rsx decode returns 0x7fffffff when restart index is 0xffffffff. For some reason bit 31 is 0 when it shouldn't be.
Added vulkan vertex textures and fix for cases where custom restart indices would be used

D3D12 vertex textures is still WIP. I'm almost certain I messed up the descriptor bindings somehow and I'm short on time to keep digging. @vlj please take a look (I get the invalid parameter error)

@raven02 please test for vulkan and ogl regressions

Some more work needs to go into the decompiler itself to get the vertex index being sampled. The dst opcode structure still contains some unknowns. I'll have to ask from those familiar with rsx internals (@vlj) or refer to nouveau docs for the ISA. If anyone has insight into this feel free to chip in. Should probably be  a 2-bit encoding in one of the Dx sections.

File: rpcs3/Emu/RSX/gcm_enums.cpp [hunk 1]
```diff
@@ -563,6 +563,7 @@ enum
 	CELL_GCM_TEXTURE_MAX_ANISO_16 = 7,
 
 	// Texture Filter
+	CELL_GCM_TEXTURE_FILTER_NONE = 0,
```
Comment (Nekotekina) [inline rpcs3/Emu/RSX/gcm_enums.cpp]: There is no such filter.

**Comment by raven02** [comment]
Tested for some games with both GL and Vulkan backend .It looks good.

File: rpcs3/Emu/RSX/gcm_enums.cpp (hunk 1 above)
Comment (kd-11) [inline rpcs3/Emu/RSX/gcm_enums.cpp]: Vertex textures do not specify filtering. In the geforce 7 era, filtering vs textures was not supported at all. I checked this, and on the vertex_texture sample, texture filter modes are set to 0.

Comment (Nekotekina) [inline rpcs3/Emu/RSX/gcm_enums.cpp]: Whatever they do, CELL_GCM_TEXTURE_FILTER_NONE simply doesn't exist. We didn't invent these constants, it's confusing. And note my recent issue about gcm_enums, this problem seems relevant that the values are checked in the situation they shouldn't.

**Comment by SakataGintokiYT** [comment]
Fotoszok.pl

**Comment by kd-11** [comment]
I'll have to revert some texture format changes on vulkan/gl.

**Comment by kd-11** [comment]
@SakataGintokiYT Are the vulkan/gl issues regressions? If so, @raven02 I'll need your help to figure out which texture format I botched here. I only changed swizzles for X32/16 and W32_Z32_Y32_X32 which were wrong based on inspection of shader code using the textures.

**Comment by SakataGintokiYT** [comment]
@kd-11 
No , is not a regression on Opengl and Vulkan it was so all the time

**Comment by SakataGintokiYT** [comment]
@kd-11 
This is not a regression

**Comment by kd-11** [comment]
Minor bugfixes and reworked a few areas.
Currently:

OpenGL
image

Vulkan
image

For D3D12, the shaders generated are okay, but the upload/binding code is probably where it breaks.

**Comment by raven02** [comment]
@SakataGintokiYT , is master also broken ? as i found some games already broken with master and not related to this vertex texture enhancement.

**Comment by SakataGintokiYT** [comment]
@kd-11 
@raven02 
The games of naruto are broken in master build , in VK and GL Is not a regression this commit

**Comment by raven02** [comment]
Thanks for checking this out .It would be nice if you can cross check which commit start the breaking so we can fix the regression .

**Comment by SakataGintokiYT** [comment]
@raven02 
Naruto games only work in DX12  .  Never work in Vulkan and OGL so it's not a regression

**Comment by kd-11** [comment]
I've removed dx12 part for now. The shaders and upload code is fine, but the resource binding/descriptor stuff is causing a mismatch. See https://github.com/kd-11/rpcs3/tree/dx12

Travis-Ci and appveyor are crashing before compilation starts, I'm not sure why. Should be good to go as part 1 for now as it shouldn't cause any regressions.

**Comment by raven02** [comment]
Need someone to restart appveyor

**Comment by kd-11** [comment]
Appveyor failing again during pre-compilation setup

**Comment by raven02** [comment]
@Nekotekina , please help to restart the appveyor.

**Pull request merged**